### ISSUE TITLE
Fix crashes and obscure errors in tests when enabling --exact-data-cons

### ISSUE THREAD

**Issue by facundominguez** [ucsd-progsys/liquidhaskell#2292]
#2290 tried enabling `--exact-data-cons` by default, but this causes many tests to fail with crashes and error messages that are hard to relate to the flag. While the feature could still be disabled in the tests, we don't want users to be confronted with the problem of discovering how to get rid of such errors.

This issue is about improving those error messages.

Earlier discussion on enabling `--exactdc` by default can be found in https://github.com/ucsd-progsys/liquidhaskell/issues/1284.

------
Work in progress: https://github.com/ucsd-progsys/liquidhaskell/compare/fd/exactdc

**Comment by ranjitjhala** [comment]
I agree - “exact” should be the default. The trouble is that some things cannot be translated to SMT eg if you have function types in the data definition.. while others can… eg plain Lists …

**Comment by facundominguez** [comment]
> eg if you have function types in the data definition.. 

IIRC, functions are replaced with integers to describe higher order functions to the SMT solver. Could we hope to do the same with data constructors?

**Comment by facundominguez** [comment]
I started a discussion in #1284 to try to pinpoint field types that could not be used with `--exact-data-cons`.

**Comment by facundominguez** [comment]
At the moment I have a few errors similar to 
```
Sort Error in Refinement: {VV : (Tuple3 a##49_1 b##49_2 c##49_3) | is$GHC.Internal.Tuple.(,,) VV
      && lqdc##$select##GHC.Internal.Tuple.(,,)##1 VV == x_Tuple31##GHC.Internal.Tuple.(,,)
      && lqdc##$select##GHC.Internal.Tuple.(,,)##2 VV == x_Tuple32##GHC.Internal.Tuple.(,,)
      && lqdc##$select##GHC.Internal.Tuple.(,,)##3 VV == x_Tuple33##GHC.Internal.Tuple.(,,)
      && x_Tuple31 VV == x_Tuple31##GHC.Internal.Tuple.(,,)
      && x_Tuple32 VV == x_Tuple32##GHC.Internal.Tuple.(,,)
      && x_Tuple33 VV == x_Tuple33##GHC.Internal.Tuple.(,,)}
Expressions is$GHC.Internal.Tuple.(,,) VV should have bool sort, but has GHC.Internal.Types.Bool                                                                                                  
Just constructor 
```
in `src/Data/Bits_LHAssumptions.hs`.

These errors are because `--exact-data-cons` creates measures for wired data constructors anew in every module where it is enabled. In the case of the `liquidhaskell` package, the embed annotations that would make `bool` and `GHC.Internal.Types.Bool` compatible are not yet in effect.

These errors are easy enough to avoid with:
```diff
--- a/liquidhaskell.cabal
+++ b/liquidhaskell.cabal
@@ -84,7 +84,7 @@ library
                       containers,
                       ghc-prim
   default-language:   Haskell98
-  ghc-options:        -Wall
+  ghc-options:        -Wall -fplugin-opt=LiquidHaskellBoot:--exact-data-cons=False
 
   if flag(devel)
     ghc-options:      -Werror
```

Other modules then fail when using `{-# LANGUAGE NoImplicitPrelude #-}`. If no Prelude is imported, the embed annotations for wired data constructors are not available either.

A solution to this could be to always enable the (effect of the) embed annotation for Bool. **I implemented this in the work in progress.**

**Comment by facundominguez** [comment]
There is another error in `tests/pos/LambdaEvalTiny.hs`

```
...
data Expr
  = Lam Bndr Expr
  | Var Bndr
  | App Expr Expr

{-@
data Expr
  = Lam { eX :: Bndr, eBody :: Expr }
  | Var { eX :: Bndr                }
  | App { eF :: Expr, eArg  ::Expr  }
@-}
...
```

The error is about a duplicated definition of `eX` in the SMT solver.

<details>

```
<no location info>: error:
    HasCallStack backtrace:
  collectExceptionAnnotation, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:170:37 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:90:42 in ghc-internal:GHC.Internal.Exception
  throw, called at src/Language/Fixpoint/Types/Errors.hs:151:7 in liquid-fixpoint-0.9.6.3.5-inplace:Language.Fixpoint.Types.Errors
  die, called at src/Language/Fixpoint/Smt/Interface.hs:496:19 in liquid-fixpoint-0.9.6.3.5-inplace:Language.Fixpoint.Smt.Interface
  respSat, called at src/Language/Fixpoint/Smt/Interface.hs:465:17 in liquid-fixpoint-0.9.6.3.5-inplace:Language.Fixpoint.Smt.Interface
  smtCheckUnsat, called at src/Language/Fixpoint/Solver/Monad.hs:194:16 in liquid-fixpoint-0.9.6.3.5-inplace:Language.Fixpoint.Solver.Monad

:1:1-1:1: Error
  crash: SMTLIB2 respSat = Error "line 3 column 447: invalid datatype declaration, repeated accessor identifier 'LambdaEvalTiny.eX'"
```
</details>

The issue here seems to be that `--exact-data-cons` causes LH to send a datatype declaration to the SMT solver, and the SMT solver isn't fond of field names being reused on different constructors. Perhaps we can keep `--exact-data-cons` for deciding whether to send data declarations to the SMT solver.

---------

There is an odd case related to the above in `tests/pos/OrdList.hs`:
```
{-@
data OrdList [ollen] a = None
                      | One  (ox  :: a)
                      | Many (oxs1 :: ListNE a)
                      | Cons (ox  :: a)           (oxs3 :: OrdList a)
                      | Snoc (oxs2 :: OrdList a)  (ox  :: a)
                      | Two  (ox  :: OrdListNE a) (oy  :: OrdListNE a)
@-}
```
What should be the sort of the selector `ox`?

At the moment, without `--exact-data-cons` it is taken to be `ox :: OrdList a -> OrdList a`. With `--exact-data-cons`, it is taken to be `ox :: OrdList a -> a` which causes LH to reject the expressions that use it as ill-sorted.

It is looking to me like this data annotation should be rejected. I don't think we can allow a selector to have different sorts depending on the data constructor in which it appears. **I implemented this in the work in progress.**

**Comment by facundominguez** [comment]
One last hindrance is that `--exact-data-cons` decides what selectors to produce based on the types that are used in reflected functions or measures.

This is problematic when we want to reflect or make a measure of a function that uses a type with unpacked fields like `ByteString`.

```
{-@ LIQUID "--exact-data-cons" @-}                                                                                                                                                                  
module Test where
import qualified Data.ByteString as BS
{-@ reflect bsLen @-}
bsLen :: [BS.ByteString] -> Int
bsLen xs = length xs
```

An error complains that the data constructor has different arity than expected because the `Data.ByteString` module is built with optimizations, and the fields inside `ByteString` use the `UNPACK` pragma.

Ideally, we would only produce selectors for data constructors that are actually used. The `bsLen` example above does not need selectors for the fields of `ByteString`. **I also implemented this in the work in progress**
